documentation/c11-xmlproc.sil

```diff
@@ -181,7 +181,7 @@ any counts we have for levels above the current level (if we didn’t do that, t
 Now we find the title, and prefix it by the concatenation of all the \code{seccount}s:
 
 \begin{verbatim}
-  local title = SILE.findInTree(content, "title")
+  local title = SILE.inputter:findInTree(content, "title")
```

Comment:
Hey @alerque ...
Commenting on a random line of this huge PR, because I am getting very very lost here.

But `findInTree` shouldn't IMHO be be in an "inputter" scope, because it doesn't depend on it. It's an AST-level function. It works regardless of the inputter, at internal SILE AST level (i.e. _after_ any input was read, from any format); It is, somehow, on the same level as `contentToString` etc. -- the internal SILE AST, independent from the input format.

(some format) --> inputter --> (internal AST)

The internal AST being what `SILE.process()` etc. act upon. This "findInTree" is in the latter, no?

Reply:
>  It works regardless of the inputter,

I am using it, actually, in commands. Regardless they came from XML, SIL, or Markdown.

Reply:
> But findInTree shouldn't IMHO be be in an "inputter" scope, because it doesn't depend on it. It's an AST-level function.

It's not in any *specific* inputter, it is in the base inputter class so it is available in any inputter. And yes it functions on the AST, but we don't have an AST before we parse and we might want to walk the AST before we process so it isn't that illogical a place for it. A case could be made for putting it in utilities, but I didn't move it here with any recent refactoring, it was already here and I just made it accessible to all the inputters instead of just one.

Reply:
I agree to somehow disagree. It's used on AST = it doesn't even depend on a "base inputter', it's not an inputter-related method. It's on the same level, as I said, as some other methods _that shouldn't even be on a "utilities" thing_, but on an AST-related helper. There's a subtle difference, but there is one, no?

Reply:
Sure, of course I see more than a subtle difference.

Actually what needs to happen in the long term is the AST needs to be a class of it's own with it's own methods, that way any AST fragment could be handled with the same tooling and the tooling would follow the object. This will make it much more portable later when/if the AST we adjust the data model or want to pass it off somewhere.

My only point is that I didn't just put it under inputters willy-nilly, it was already there and I just refactored the inputters in a way that wouldn't break too much at once but would make room for the inputters being modules that could be added to easily.

Reply:
My viewpoint: move it to SILE.ast.findInTree. Move contentToString etc. there too = Everything we currently have that's related to the AST structure. Whether the AST is a class (with only one instance? Oh well) _eventually_ doesn't matter yet. Add SILE.ast.extratFromTree (lol, one of my very first issues back in Autumn 2021, and I keep re-implementing it in packages needing it...). Don't entangle things unrelated to inputters in them.

Reply:
Sure, that sounds good.

There would quite likely me many instances of an AST class. Each file read (master, prelude, etc.), snippet parsed from a different language (sil in xml or xml in sil or string parsed as sil from lua or whatever) would be an AST instance. Likewise for each `content` object passed to child functions: they would be instances of an AST with the remaining leaves than need processing.

My next priority is to get the typesetter object untangled a bit because that's where a lot of the bugs affecting my publications are, but getting AST handling organized is on my list already.

Reply:
> ... would be an AST instance. 
Ponder that, though: even Pandoc has one AST for all the input formats it eventually handles. The AST is the centra pivot. I do have some issues with the SILE AST, but that may be raised later (e.g. we have position on commands, but not on text nodes... which are just strings... Let's not break there yet, though, Simon's AST is simple but efficient. IMHO we should only have one AST. Inputters are responsible to convert whatever format into it.

Reply:
I don't think we're speaking the same language. I didn't mean there should be different ASTs — in fact I would think the SILE core would only have a single AST implementation as a single class: but in the course of processing we might have more than one *instance* of the AST class.

> Inputters are responsible to convert whatever format into it.

Of course, but we have more than one input in the course of processing so we would come up with multiple ASTs. A major feature of a class that handles this would be `:splice()` and `:split()` methods.

Reply:
Closing this remark. (While interesting, it's not related anyway).